RFC: Ethereum attestation (#557)
radicle-dev/radicle-link: docs/rfc/ethereum_attestation.md

# RFC: Ethereum attestation

* Author: @CodeSandwich
* Date: 2021-03-05
* Status: draft
* Community discussion: n/a

## Motivation

The attestation between Link and Ethereum is a valuable building block for a user identity.
It brings the Link reputation coming from projects and contributions to the

> Ethereum world of DAOs and donations, where it's important to know, who's behind an address.

Reviewer: ```suggestion
Ethereum world of DAOs and donations, where it's important to know who's behind an address.
```

Author: Fixed

> On the other hand it lends Ethereum account reputation with its assets and undeniable history

Reviewer: ```suggestion
On the other hand, it lends Ethereum account reputation with its assets and undeniable history
```

Author: Fixed

to Link to build user's trust in the identity.

## Overview

This RFC is built on top of [Identity Proofs RFC][rfc].
It introduces support for Ethereum address claims on Link
and a smart contract on Ethereum to make Link identity claims on Ethereum.

## Link identity JSON extension

The identity JSON supports a new key: `ethereum`.
Under this key there is stored an ethereum address claim following this convention:

- `account` - the claimed ethereum address, encoded according to [EIP-55][eip-55],
e.g. using [ethers.js][ethers-addr]

> - `proof` - always `null`

Reviewer 2: This is optional, so if we aren't using it, it shouldn't be set?

Author: Fixed

Example:
```json
{
    "ethereum": {
        "account": "0xAb5801a7D398351b8bE11C439e05C5B3259aeC9B",
        "expiration": {
            "created": 1614768373,
            "expires": 31536000
        },
        "proof": null
    }
}
```

## Ethereum smart contract

> A new smart contract is deployed to the network, which is used to publish link claims:

Reviewer: This sounds a little to vague to me. I'd recommend be more specific and say something like:

```suggestion
A new Ethereum smart contract is deployed to the network, which allows users to claim their Radicle ID.
```

```solidity
contract Claims {
    event Claimed(address indexed addr);
    function claim(uint8 version, uint256 id) public;
}
```

In order to claim an ID, call `claim` using your Ethereum account.
It will emit an event `Claimed`, which later can be queried to discover your attestation.

> The claims have no expiration date and don't need to be renewed.

Reviewer 2: This means that the link can be revoked with only the ethereum account, right? Is that desirable? Basically, since the Link identity doesn't point to a specific attestation, it can be revoked from the ethereum side, without having the Link key.

Author: That's right, if you loose your Link key, you need to be able to revoke the attestation using only your Ethereum key.

Reviewer 3: The payload above has an expiration date. So this "claim" seems to be false :smirk:

Every new claim invalidates previous ones made with the same account.
To only revoke a claim without creating a new one, use ID `0`,
which is guaranteed to not match any existing identity.

Currently supported `version` values:
- `1` - an `id` is a SHA-1 root hash. The excess high bytes are zeros, e.g. for hash
`fb3102b74d7254eed7f18a31a3ba1ea946bb1a99` the passed `id` is
`000000000000000000000000fb3102b74d7254eed7f18a31a3ba1ea946bb1a99`
- `2` - an `id` is a SHA-256 root hash

Everybody must use the same smart contract address.

> Claims sent to an address not known by the users won't be noticed by anyone and will be useless.

Reviewer: > an address not known by the users won't be noticed by anyone and will be useless.

Could rephrase to in a way so that it transmits the idea of there being an official contract address that is the only one taken into account?

Author: Fixed

For reference the signature payload content is listed [here][rpc-sign].
- Read the link ID from the transaction data
- Verify that the link ID claims back the Ethereum address,
see [Discovery from a Link ID](#discovery-from-a-link-id)

## Discovery from a Link ID

When you have a link ID, you can find the claimed Ethereum address.
Obtain the tip of its identity chain and read the ethereum address from the identity JSON
`account` field in section `ethereum`, unless it's expired.
You need to verify that the given Ethereum address claims back the link ID,
see [Discovery from an Ethereum address](#discovery-from-an-ethereum-address).

## Revocation of an attestation

When your attestation for whatever reason is no longer valid,
you should revoke it as soon as possible.
Only one claim needs to be revoked to break the attestation,
but to improve security you should revoke both sides if you can.

To revoke a claim on Link, update and publish the identity JSON.
You can change the claimed Ethereum address or remove the `ethereum` section altogether.

To revoke a claim on Ethereum, call the `claim` function in `Claims` contract.

> You can claim a different link ID or an ID `0` to stop claiming anything.

Reviewer: > to stop claiming anything.

could we rephrase so something like:

"to revoke any claim you may have."

Author: Fixed

# RFC: Ethereum attestation

* Author: @CodeSandwich
* Date: 2021-03-05
* Status: draft
* Community discussion: n/a

## Motivation

The attestation between Link and Ethereum is a valuable building block for a user identity.
It brings the Link reputation coming from projects and contributions to the
Ethereum world of DAOs and donations, where it's important to know who's behind an address.
On the other hand, it lends Ethereum account reputation with its assets and undeniable history
to Link to build user's trust in the identity.

## Overview

> This RFC is built on top of [identities spec][identities].

Reviewer 3: ```suggestion
This RFC is built on top of the [identities spec][identities].
```

Author: Fixed

- `address` - the claimed ethereum address, encoded according to [EIP-55][eip-55],
e.g. using [ethers.js][ethers-addr]
- `expiration` - the claim expiration timestamp, encoded as a [JavaScript Date][date]

Example:
```json
{
    "payload": {
        "https://radicle.xyz/ethereum/claim/v1": {
            "address": "0xAb5801a7D398351b8bE11C439e05C5B3259aeC9B",
            "expiration": "2021-03-19T23:15:30.000Z"
        }
        ...
    }
    ...
}
```

## Ethereum smart contract

A new Ethereum smart contract is deployed to the network, which lets users claim their Radicle IDs:

```solidity
contract Claims {
    event Claimed(address indexed addr);
    function claim(uint8 version, uint256 id) public;
}
```

To claim an ID, call `claim` using your Ethereum account.
It will emit an event `Claimed`, which later can be queried to discover your attestation.
The claims have no expiration date and don't need to be renewed.

Every new claim invalidates previous ones made with the same account.
To only revoke a claim without creating a new one, use ID `0`,
which is guaranteed to not match any existing identity.

> Currently supported `version` values:

Reviewer 3: Is `version` supposed to be in the payload above? Or where did it come from?

Reviewer 3: Or would you use `claim/v1` and `claim/v2` URLs?

Author: The payload is the Link-stored data, the `version` is used only in Ethereum. It has no corelation with the Link payload URL versioning.

Reviewer 3: Ah, I missed it in the contract function :)

- `1` - an `id` is a SHA-1 root hash. The excess high bytes are zeros, e.g. for hash
`fb3102b74d7254eed7f18a31a3ba1ea946bb1a99` the passed `id` is
`000000000000000000000000fb3102b74d7254eed7f18a31a3ba1ea946bb1a99`
- `2` - an `id` is a SHA-256 root hash

> We need to deploy an official instance of the `Claims` smart contract and

Reviewer 3: How do we attest this? weh weh weh

Author: We don't, there can be multiple instances deployed and it won't affect the overall security. If you start using a different instance, you probably won't notice other people's claims and they won't notice yours

The example calls are based on the standard [client JSON RPC API][rpc] and should be exposed
by your favourite Ethereum client library.
It's important that the client must be trusted not to hide the events.

- Use [getLogs][rpc-logs] to get the newest `Claimed` event filtered for the given ethereum address
- Get the event's `transactionHash` field and use it to fetch the transaction which emitted it with
[getTransactionByHash][rpc-tx]
- Validate that the transaction signature matches its data and the ethereum address.
For reference the signature payload content is listed [here][rpc-sign].
- Read the link ID from the transaction data
- Verify that the link ID claims back the Ethereum address,
see [Discovery from a Link ID](#discovery-from-a-link-id)

## Discovery from a Link ID

When you have a link ID, you can find the claimed Ethereum address.
Obtain the tip of its identity chain and read the ethereum address from the identity JSON
`address` field in section `https://radicle.xyz/ethereum/claim/v1`, unless it's expired.
You need to verify that the given Ethereum address claims back the link ID,
see [Discovery from an Ethereum address](#discovery-from-an-ethereum-address).

> ## Revocation of an attestation

Reviewer 3: It might be good to call out here that a revocation will only be as good as the propagation of the document in the network. What I mean is that if I revoke my address, I still have to wait for others to fetch those changes -- so _could_ still end up transferring MONIEZ to the old address.

Author: Added

## Overview

This RFC is built on top of [identities spec][identities].
It introduces support for Ethereum address claims on Link
and a smart contract on Ethereum to make Link identity claims on Ethereum.

## Link identity JSON extension

The identity JSON `payload` structure supports a new key: `https://radicle.xyz/ethereum/claim/v1`.
Under this key there is stored an ethereum address claim following this convention:

- `address` - the claimed ethereum address, encoded according to [EIP-55][eip-55],
e.g. using [ethers.js][ethers-addr]
- `expiration` - the claim expiration timestamp, encoded as a [JavaScript Date][date]

Example:
```json
{
    "payload": {
        "https://radicle.xyz/ethereum/claim/v1": {

>             "account": "0xAb5801a7D398351b8bE11C439e05C5B3259aeC9B",

Reviewer: s/account/address

Author: Fixed

When you have a link ID, you can find the claimed Ethereum address.
Obtain the tip of its identity chain and read the ethereum address from the identity JSON

> `account` field in section `https://radicle.xyz/ethereum/claim/v1`, unless it's expired.

Reviewer: s/account/address

Author: Fixed
